app/src/main/java/com/kickstarter/viewmodels/ChangePasswordViewModel.kt

```diff
@@ -1,165 +1,69 @@
-            return this.apolloClient.updateUserPassword(changePassword.currentPassword, changePassword.newPassword, changePassword.confirmPassword)
-                .doOnSubscribe { this.progressBarIsVisible.onNext(true) }
-                .doAfterTerminate { this.progressBarIsVisible.onNext(false) }
-        }
-
-        fun updatePasswordData(oldPassword: String, newPassword: String) {
-            this.currentPassword.onNext(oldPassword)
-            this.newPassword.onNext(newPassword)
-            this.confirmPassword.onNext(newPassword)
-        }
-
-        fun resetError() {
-            this.error.onNext("")
-        }
-
-        override fun changePasswordClicked() {
-            this.changePasswordClicked.onNext(Unit)
-        }
-
-        override fun confirmPassword(confirmPassword: String) {
-            this.confirmPassword.onNext(confirmPassword)
-        }
-
-        override fun currentPassword(currentPassword: String) {
-            this.currentPassword.onNext(currentPassword)
-        }
-
-        override fun newPassword(newPassword: String) {
-            this.newPassword.onNext(newPassword)
-        }
-
-        override fun error(): Observable<String> {
-            return this.error
-        }
-
-        override fun progressBarIsVisible(): Observable<Boolean> {
-            return this.progressBarIsVisible
-        }
-
-        override fun success(): Observable<String> {
-            return this.success
-        }
-
-        data class ChangePassword(val currentPassword: String, val newPassword: String, val confirmPassword: String) {
-            fun isValid(): Boolean {
-                return this.currentPassword.validPassword() &&
-                    this.newPassword.validPassword() &&
-                    this.confirmPassword.validPassword() &&
-                    this.confirmPassword == this.newPassword
-            }
-
-            fun warning(): Int =
-                newPassword.newPasswordValidationWarnings(confirmPassword) ?: 0
+                .catch {
+                    mutableUIState.emit(UpdatePasswordUIState(errorMessage = it.message ?: "", isLoading = false))
+                }
+                .collect()
         }
+    }
 
-        override fun onCleared() {
-            disposables.clear()
-            super.onCleared()
+    fun resetError() {
+        viewModelScope.launch {
+            mutableUIState.emit(UpdatePasswordUIState(errorMessage = null))
         }
     }
+}
 
-    class Factory(private val environment: Environment) : ViewModelProvider.Factory {
-        override fun <T : ViewModel> create(modelClass: Class<T>): T {
-            return ChangePasswordViewModel(environment) as T
-        }
+class ChangePasswordViewModelFactory(private val environment: Environment) :
+    ViewModelProvider.Factory {
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
+        return ChangePasswordViewModel(environment) as T
```

Comment:
Just seeing a class go down ~100 lines while maintaining functionality really excites me for coroutines and the UI state model

Reply:
especially since this is a smaller use case